docs/raw/push-notification-server.md

```diff
@@ -0,0 +1,765 @@
+- The hash of the public keys the user is interested in
+
+### PushNotificationQueryInfo
+
+`access_token`: the access token used to send a push notification 
+`installation_id`: the `installation_id` of the device associated with the `access_token`
+
+### PushNotificationQueryResponse
+
+`info`:  a list of `PushNotificationQueryInfo`
+
+### PushNotification
+
+`access_token`: the access token used to send a push notification
+`chat_id`: the `SHA3-256` of the `chat_id`
+
+### Data disclosed
+
+- The `SHA3-256` of the `chat_id` the notification is to be sent for
+
+### PushNotificationRequest
+
+`requests`: a list of `PushNotification`
+`message`: the encrypted message that we want to notify on
+`message_id`: the [status message id](./6-payloads.md)
+`ack_required`: a 32 bytes long AES key, only set if an ack is required
+
+### Data disclosed
+
+- The status message id for which the notification is for
+- the cypher text of the message
+
+### PushNotificationAcknowledgement
+
+`id`: the `ack_required` string passed by the client
+
+## Partitioned topic
+
+This is a modification of the [partitioned topic](./10-waku-usage.md#partitioned-topic) used by clients
+
+```golang
+var partitionsNum *big.Int = big.NewInt(5000)
+var partition *big.Int = big.NewInt(0).Mod(publicKey.X, partitionsNum)
+
+partitionTopic := "push-notifications-" + strconv.FormatInt(partition.Int64(), 10)
+
+var hash []byte = keccak256(partitionTopic)
+var topicLen int = 4
+
+if len(hash) < topicLen {
+    topicLen = len(hash)
+}
+
+var topic [4]byte
+for i = 0; i < topicLen; i++ {
+    topic[i] = hash[i]
+}
+```
+
+
+## Anonymous mode of operations
+
+An anonymous mode of operations MAY be provided by the client, where the 
+responsibility of propagating information about the user is left to the client,
+in order to preserve privacy.
+
+A client in anonymous mode can register with the server using a key different 
+from their identity key.
+This will hide their real identity key.
+
+This identity key is effectively a secret and SHOULD only be disclosed to clients that you the user wants to be notified by.
+
+A client MAY advertise the access token on the contact-code topic of the key generated.
+A client MAY share their identity key through [contact updates](./6-payloads.md#contact-update)
+
+A client receiving a push notification identity key SHOULD listen to the contact code 
+topic of the push notification identity key for updates.
+
+If a client has received a push notification identity key they MAY request 
+the access token using an ephemeral key when querying the server.
+
+The method described above effectively does not share the identity of the sender
+nor the receiver to the server, but MAY result in missing push notifications as
+the propagation of the secret is left to the client.
+
+This can be mitigated by [device syncing](./6-payloads.md), but not completely
+addressed.
+
+[//]: This section can be removed, for now leaving it here in order to help with the 
```

Comment:
The  whole thing or only the device syncing part? I can further elaborate if necessary

Reply:
Ok, I think I understand this now (except device syncing). The terminology (specifically, use of the words identity key, push notification identity key and secret :) threw me off earlier, sorry. 

So the difference is in the generation of a push notification key (PN key) which is different from the user's chat key (i.e. identity key). The msg sender has to be aware of this PN key to send push notifications to the recipient and the probabilistic (?) propagation of this PN key makes the receipt of PNs indeterministic. 

The PN sender may get the access token either by querying the PN server like earlier (using an ephemeral key if it wants) or directly receive it from the client by listening to contact-code topic for the PN key. The PN sender then send a notification just like earlier by talking to the PN server which in turn talks to the gorush server.

The use of ephemeral PN-specific keys makes this mode anonymous.

Reply:
Correct,
the device syncing part is that if you have multiple devices, say A1 & A2, and you chose to propagate the token yourself (so not through the server), you need to make sure that A1 & A2 are paired, otherwise they might not be aware of each other, and therefore you might be propagating only the token for A1 to another client for example, which will only send a PN to A1.